vamb/__main__.py

```diff
@@ -939,6 +980,306 @@ def run(
+    df_mmseq = df_mmseq[~(df_mmseq[2] == 'no rank')]
+
+    ind_map = {c: i for i, c in enumerate(contignames)}
+    indices_mmseq = [ind_map[c] for c in df_mmseq[0]]
+
+    graph_column = df_mmseq[8]
+    nodes, ind_nodes, table_parent = vamb.h_loss.make_graph(graph_column.unique())
+
+    classes_order = np.array(list(graph_column.str.split(';').str[-1]))
+    targets = [ind_nodes[i] for i in classes_order]
+
+    model = vamb.h_loss.VAMB2Label(
+        rpkms.shape[1], 
+        len(nodes), 
+        nodes, 
+        table_parent,
+        cuda=vamb_options.cuda,
+    )
+
+    dataloader_vamb, mask_vamb = vamb.encode.make_dataloader(rpkms, tnfs, lengths, batchsize=predictor_training_options.batchsize)
+    dataloader_joint, mask_joint = vamb.h_loss.make_dataloader_concat_hloss(rpkms[indices_mmseq], tnfs[indices_mmseq], lengths[indices_mmseq], targets, len(nodes), table_parent, batchsize=predictor_training_options.batchsize)
+
+    print("", file=logfile)
+    log("Created dataloader and mask", logfile, 0)
+    n_discarded = len(mask_vamb) - mask_vamb.sum()
+    log(f"Number of sequences unsuitable for encoding: {n_discarded}", logfile, 1)
+    log(f"Number of sequences remaining: {len(mask_vamb) - n_discarded}", logfile, 1)
+
+    composition.metadata.filter_mask(mask_vamb)
+    names = composition.metadata.identifiers
+
+    predictortime = time.time()
+    log(f"Starting training the taxonomy predictor at {str(datetime.datetime.now())}", logfile, 0)
+    MODEL_PATH = vamb_options.out_dir.joinpath("predictor_model.pt")
+    with open(MODEL_PATH, 'wb') as modelfile:
+        model.trainmodel(
+            dataloader_joint,
+            nepochs=predictor_training_options.nepochs,
+            modelfile=modelfile,
+            logfile=logfile,
+            batchsteps=predictor_training_options.batchsteps,
+        )
+    log(f"Finished training the taxonomy predictor in {round(time.time() - predictortime, 2)} seconds.", logfile, 0)
+    predicted_vector = model.predict(dataloader_vamb)
+
+    log(f"Writing the taxonomy predictions", logfile, 0)
+    df_gt = pd.DataFrame({'contigs': names})
+    nodes_ar = np.array(nodes)
+    predictions = []
+    for i in range(len(df_gt)):
+        predictions.append(';'.join(nodes_ar[predicted_vector[i] > 0.5][1:]))
+    df_gt[f'predictions'] = predictions
+
+    df_mmseq_sp = df_mmseq[(df_mmseq[2] == 'species')]
+    mmseq_map = {k: v for k, v in zip(df_mmseq_sp[0], df_mmseq_sp[8])}
+    preds = []
+    for i, r in df_gt.iterrows():
+        pred_line = r[f'predictions'].split(';')
+        try:
+            mmseq_line = mmseq_map.get(r['contigs'], '').split(';')
+        except AttributeError:
+            preds.append(';'.join(pred_line))
+            continue
+        if mmseq_line[0] != '':
+            for i in range(len(mmseq_line)):
+                if i < len(pred_line):
+                    pred_line[i] = mmseq_line[i]
+        preds.append(';'.join(pred_line))
```

Comment:
I don't really understand the logic here. That might be OK, but perhaps a comment would be in order.